File: sdk/android/signallake-android/src/main/java/dev/signallake/BatchBuilder.java
package dev.signallake;

import java.util.List;

final class BatchBuilder {
    private BatchBuilder() {}

    static EventBatch buildBatch(String batchId, String createdAt, Source source, List<EventEnvelope> events) {
        return new EventBatch(batchId, createdAt, source, events);
    }

    static EncryptedEventBatch buildEncryptedBatchEnvelope(
            EventBatch batch,
            String keyId,
            String nonce,
            String ciphertext,
            String authTag) {
        return new EncryptedEventBatch(
                batch.batchId,
                batch.createdAt,
                new UploadSource(
                        batch.source.appId,
                        batch.source.product,
                        batch.source.sdkName,
                        batch.source.sdkVersion,
                        batch.source.environment),
                new PlaintextDescriptor(EventBatch.SCHEMA_VERSION, EventBatch.CONTENT_TYPE),
                new EncryptionMetadata(EncryptedEventBatch.ALG_AES_256_GCM, keyId, nonce),
                new EncryptedPayload("base64url", ciphertext, authTag));
    }
}

File: sdk/js/core/src/batch-builder.mjs
import { defaultId, defaultNow, normalizeIsoDate } from "./ids.mjs";

export function buildBatch(options) {
  const source = options?.source;
  const events = options?.events ?? [];
  const id = options?.id ?? defaultId;
  const now = options?.now ?? defaultNow;

  if (!source || typeof source !== "object") {
    throw new TypeError("source must be an object");
  }
  if (!Array.isArray(events) || events.length === 0) {
    throw new TypeError("events must be a non-empty array");
  }

  return {
    schemaVersion: "signallake.event-batch.v1",
    batchId: options.batchId ?? id(),
    createdAt: normalizeIsoDate(options.createdAt ?? now(), "createdAt"),
    source: {
      appId: source.appId,
      product: source.product,
      sdkName: source.sdkName,
      sdkVersion: source.sdkVersion,
      environment: source.environment
    },
    eventCount: events.length,
    compression: "none",
    events: events.map((event) => structuredClone(event))
  };
}
